### ISSUE TITLE
Registration by Convention?

### ISSUE THREAD

**Issue by houseofcat** [ipjohnson/Grace#232]
I am really looking to give Grace a try on a NetCore 3.0 project.

I see you can find named instance (withKey:) lookups, which was a requirement for us.

The last thing we were looking at was RegistrationByConvention (ala Unity)?

AllClasses from a specific assembly, with matching name interfaces (MyClass -> IMyClass).

     var repositoriesAndServices =
         AllClasses
        .FromLoadedAssemblies()
        .Where(obj => obj.ToString().Contains("MyProject.Services.")
             || obj.ToString().Contains("MyProject.Repositories."));

     container.RegisterTypes(repositoriesAndServices,
         WithMappings.FromMatchingInterface,
         WithName.Default,
         WithLifetime.ContainerControlled);

I see another ticket #136 that mentions Strategies.  If this is indeed how I would go about using resolution strategies, could you provide a demonstration? I didn't find this in the Wiki/Api Docs.

**Comment by ipjohnson** [comment]
Hi @houseofcat 

I'm not all that familiar with unity bulk registration so I don't recognize what the With* parameters really represent.

That said Grace does support bulk registration by interface & keyed interface. It all is built off of the `Export(IEnumerable<Type>)` method. 

If you can explain a bit more what you're specifically looking to do I can direct you a little better.

**Comment by silkfire** [comment]
I think he's looking for something like default conventions, similar to how StructureMap does it:

https://structuremap.github.io/registration/auto-registration-and-conventions/  
see paragraph titled _The Default ISomething/Something Convention_

**Comment by houseofcat** [comment]
Yes, something automatically add all Classes with Interface matching ClassName, i.e. IClassName.

Otherwise it is register each class and interface manually.

Nothing really fancy but can understand if it doesn't exist.

**Comment by ipjohnson** [comment]
@houseofcat I'm in the middle of doing 7.1.0 so it would be a good time to add functionality like this. 

There is already the ability to export a type set by interface or attributed interface. So adding something that exports when the interface name matches seems like a very nice convention. 

You mention the need for looking up by key. Is that a function of the fact the unity required it when registering or is this something that is located later with a key?

**Comment by houseofcat** [comment]
Thats totally awesome. Appreciate it.

The key lookup, is really, from Unity world is a named instance.

For example, I have written a custom ORM, it has type IDataSource, the DataSource wraps everything around a logical DataBase and Cache-aside layer.

But I have two DataSources, one for main and one for logging, named "main" and "logging" thusly. I just need a way to acquire a specific singleton of DataSource based on some parameter.

I assume Key registration was the Grace equivalent to Unity named instance pattern registration.

**Comment by ipjohnson** [comment]
Currently how does a class signify which data source it needs? Is it based on parameter name, attributes, or some other conventions?

**Comment by GabrielBBF** [comment]
Grace is so concise and flexible in its syntax; you can do something like this:

```
public void ConfigureContainer(IInjectionScope scope)
{
   scope.Configure(c =>
   {
      c.ExportAsKeyed<DataSource, IDataSource>("main").Lifestyle.Singleton();
      c.ExportAsKeyed<DataSourceLogging, IDataSource>("log").Lifestyle.Singleton();

      c.Export<ORM>().WithCtorParam<IDataSource>().LocateWithKey("main").As<IRepository>().Lifestyle.Singleton();
   }

// ...

}
```

**Comment by silkfire** [comment]
Grace is so concise and flexible in its syntax; you can do something like this:

```
public void ConfigureContainer(IInjectionScope scope)
{
   scope.Configure(c =>
   {
      c.ExportAsKeyed<DataSource, IDataSource>("main").Lifestyle.Singleton();
      c.ExportAsKeyed<DataSourceLogging, IDataSource>("log").Lifestyle.Singleton();

      c.Export<ORM>().WithCtorParam<IDataSource>().LocateWithKey("main").As<IRepository>().Lifestyle.Singleton();
   }

// ...

}
```

**Comment by houseofcat** [comment]
> Currently how does a class signify which data source it needs? Is it based on parameter name, attributes, or some other conventions?

Class A, has a dependency on IDataSource 1, IDataSource 2, it will obviously need to be specified.

Unity does this one of two ways, by Attribute

    [Dependency("main")]
    private static IDataSource MainDataSource { get; set; }

    [Dependency("log")]
    private static IDataSource LogDataSource { get; set; }

or at wire time:

    _container.RegisterInstance<IDataSource>(
         "main",
         new DataSource(params));

then followed by wiring up a class that needs both named instances

     container.RegisterInstance<IMyClass>(
          MyClass(
               container.Resolve<IDataSource>("main"),
               container.Resolve<IDataSource>("logger")));

**Comment by ipjohnson** [comment]
@houseofcat if it has just one dependency on IDataSource how do you pick the correct one? Does it default to main?

As for attributes vs. convention vs. explicit fluent configuration Grace can support all of these it really just depends on what your code looks like an what you'd like to do.

**Comment by houseofcat** [comment]
The constructor needed both, not just one. Sorry mobile at the moment.

**Comment by ipjohnson** [comment]
Totally doable I’ll see if I can find some time this evening and code up an example that doesn’t require hand wiring everything.

**Comment by ipjohnson** [comment]
@houseofcat sorry I meant to write this up last week but life seemed to get in the way.

I've added a test that shows you how to dynamically set the locate key based off of parameter name. This is also possible with attributes but I'm more of a fan of the name convention vs. having to attribute everywhere.

**Comment by houseofcat** [comment]
Thanks Ian!
